internal/bft/controller.go

```diff
@@ -595,112 +595,101 @@ func (c *Controller) sync() (viewNum uint64, seq uint64, decisions uint64) {
+	var latestDecisionSeq, latestDecisionViewNum, latestDecisionDecisions uint64
+	var latestDecisionMetadata *protos.ViewMetadata
+	if latestDecision.Proposal.Metadata == nil {
 		c.Logger.Infof("Synchronizer returned with proposal metadata nil")
-		response := c.fetchState()
-		if response == nil {
-			return 0, 0, 0
-		}
-		if response.View > 0 && response.Seq == 1 {
-			c.Logger.Infof("The collected state is with view %d and sequence %d", response.View, response.Seq)
-			newViewToSave := &protos.SavedMessage{
-				Content: &protos.SavedMessage_NewView{
-					NewView: &protos.ViewMetadata{
-						ViewId:          response.View,
-						LatestSequence:  0,
-						DecisionsInView: 0,
-					},
-				},
-			}
-			if err := c.State.Save(newViewToSave); err != nil {
-				c.Logger.Panicf("Failed to save message to state, error: %v", err)
-			}
-			c.ViewChanger.InformNewView(response.View)
-			return response.View, 1, 0
+		latestDecisionMetadata = nil
+	} else {
+		md := &protos.ViewMetadata{}
+		if err := proto.Unmarshal(latestDecision.Proposal.Metadata, md); err != nil {
+			c.Logger.Panicf("Controller was unable to unmarshal the proposal metadata returned by the Synchronizer")
 		}
-		return 0, 0, 0
+		latestDecisionSeq = md.LatestSequence
+		latestDecisionViewNum = md.ViewId
+		latestDecisionDecisions = md.DecisionsInView
+		latestDecisionMetadata = md
 	}
-	md := &protos.ViewMetadata{}
-	if err := proto.Unmarshal(decision.Proposal.Metadata, md); err != nil {
-		c.Logger.Panicf("Controller was unable to unmarshal the proposal metadata returned by the Synchronizer")
+
+	controllerSequence := c.latestSeq()
+	seqToReturn = controllerSequence + 1
+
+	controllerViewNum := c.currViewNumber
+	viewNumToReturn = controllerViewNum
+
+	if latestDecisionSeq > controllerSequence {
+		c.Logger.Infof("Synchronizer returned with sequence %d while the controller is at sequence %d", latestDecisionSeq, controllerSequence)
+		c.Logger.Debugf("Node %d is setting the checkpoint after sync returned with view %d and seq %d", c.ID, latestDecisionViewNum, latestDecisionSeq)
+		c.Checkpoint.Set(latestDecision.Proposal, latestDecision.Signatures)
+		c.verificationSequence = uint64(latestDecision.Proposal.VerificationSequence)
+		seqToReturn = latestDecisionSeq + 1
+		decisionsToReturn = latestDecisionDecisions + 1
 	}
 
-	latestSequence := c.latestSeq()
+	if latestDecisionViewNum > controllerViewNum {
+		c.Logger.Infof("Synchronizer returned with view number %d while the controller is at view number %d", latestDecisionViewNum, controllerViewNum)
+		viewNumToReturn = latestDecisionViewNum
+	}
 
-	if md.ViewId < c.currViewNumber {
-		c.Logger.Infof("Synchronizer returned with view number %d but the controller is at view number %d", md.ViewId, c.currViewNumber)
-		response := c.fetchState()
-		if response == nil {
-			c.Logger.Infof("Fetching state failed")
+	response := c.fetchState()
+	if response == nil {
+		c.Logger.Infof("Fetching state failed")
```

Comment:
Thinking out loud - why do we return 0,0,0 here, and not return with the existing known view and sequence known to us, which was stored on disk?